### PR TITLE
use a ready property to mark if it's ready or not

### PR DESCRIPTION
**Pull request by hcfw007**

### CODE REVIEW

**Comment by hcfw007** [comment]
https://github.com/wechaty/puppet-service/issues/191

**Review by huan — COMMENTED** [review]
What if a puppet has start-ed and ready-ed, then after a while, it was logout-ed?

In your code, it seems that this puppet will still in state `ready`?

**Comment by hcfw007** [comment]
That is a valid point, ready state should be reset if puppet emitted logout or reset event.

**Comment by hcfw007** [comment]
@huan ping

**Comment by huan** [comment]
Hi @hcfw007, thanks for reminding me about this PR.

I need some time to read and think about your design. 

It would be great if you can post the detailed explanation for your design in the PR so that we can speed up my understanding and make my decision faster.

Thank you very much.

**Comment by hcfw007** [comment]
Sure thing.

1. Why we need this ```ready``` property.
- some bot logic were written in ```onReady``` listener because it requires contact data loaded.
- if the puppet were running locally, we can simply call it with ```bot.on('ready', onReady)```
- however, if it's a remote puppet running wIth wechaty-puppet-service, the ready event will not be trigger again when t he bot connects a puppet which is already 'ready-ed'
  
2. My design to solve the above problem
- we add a ```isReady``` property in wechay-puppet. it was set to ```false``` when constructing the puppet, and set to ```true``` when the puppet is ready.
- as you pointed out, it should be set to ```false``` when the puppet is logged out or reseted.
- when wechaty-puppet-service connects the remote puppet, it should check the ```isReeady``` property, and if it's true, emit a ready event. This procedure is similar to the login event in wechaty-puppet-service

File: src/mixins/ready-mixin.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  }
+
+  return ReadyMixin
+}
+
+type ReadyMixin = ReturnType<typeof readyMixin>
+
+type ProtectedPropertyReadyMixin =
+  | 'readyIndicator'
```
Comment (huan) [inline src/mixins/ready-mixin.ts:58]: If you put any property in the `ProtectedPropertyReadyMixin`, then the final Puppet Interface will not include this property.

Does that you want or not?

File: src/mixins/ready-mixin.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  log,
+} from '../config.js'
+
+import type { PuppetSkeleton } from '../puppet/puppet-skeleton.js'
+
+const readyMixin = <MixinBase extends typeof PuppetSkeleton>(mixinBase: MixinBase) => {
+
+  abstract class ReadyMixin extends mixinBase {
+
+    readyIndicator: BooleanIndicator
```
Comment (huan) [inline src/mixins/ready-mixin.ts:13]: Here I change a boolean to the `BooleanIndicator`.

The beauty of this `BooleanIndicator` is that, you can use the below code to wait the desired state:

```ts
await puppet.readyIndicator.ready(true)
```

If the `readyIndicator` is not in the `true` status, then it will wait until it is set to `true` before the returned promise resolved.

File: src/mixins/ready-mixin.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    override async stop (): Promise<void> {
+      log.verbose('ReadyMixin', 'stop()')
+      this.readyIndicator.value(false)
+
+      /**
+       * Huan(202201) NOTE: super.stop() should be the last line of this method
+       *  becasue we should keep the reverse order of logic in start()
+       */
+      await super.stop()
```
Comment (huan) [inline src/mixins/ready-mixin.ts:47]: Please note that in your code the `super.stop()` is before other code, I moved it to the last.

In our mixin system, it should be put into the last, to make sure the mixin cleanup code will be executed before any of the super(parent) codes.

**Review by huan — APPROVED** [review]
LGTM

Please read my review comments and understand the code changes, then let me know what you think.

If you think everything is OK, we can merge this PR as you wish.

Thank you very much.

File: src/mixins/ready-mixin.ts (hunk 1 above)
Comment (hcfw007) [reply src/mixins/ready-mixin.ts:58]: I want ```readyIndicator``` exposed. My understood of this ```ProtectedPropertyMixin``` was wrong.

File: src/mixins/ready-mixin.ts (hunk 2 above)
Comment (hcfw007) [reply src/mixins/ready-mixin.ts:13]: that is a great feature.

File: src/mixins/ready-mixin.spec.ts [hunk 4]
```diff
@@ -1,18 +0,0 @@
-#!/usr/bin/env -S node --no-warnings --loader ts-node/esm
-
```
Comment (huan) [inline src/mixins/ready-mixin.spec.ts]: Please keep this unit test file with the unit test in it for future usage.

It's a good practice for pair every source code file with a unit test `.spec.ts` file.

Comment (hcfw007) [reply src/mixins/ready-mixin.spec.ts]: OK. I saw misc-mixin has no unit-test file and thought it was not necessary for mixins with no protected property.

Comment (huan) [reply src/mixins/ready-mixin.spec.ts]: I'd also like to accept a PR to add the unit test (even an empty one) to those source code files that have no `.spec.ts`.

Comment (hcfw007) [reply src/mixins/ready-mixin.spec.ts]: Sure thing, I'll try that when I have better understanding of puppet code.
